src/shims/sync.rs

```diff
@@ -0,0 +1,519 @@
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
+        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        if readers != 0 {
+            throw_unsup_format!(
+                "Deadlock due to write-locking a pthreads read-write lock while it is already read-locked"
+            );
+        } else if writers != 0 {
+            throw_unsup_format!(
+                "Deadlock due to write-locking a pthreads read-write lock while it is already write-locked"
+            );
+        } else {
+            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
+            Ok(0)
+        }
+    }
+
+    fn pthread_rwlock_trywrlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
+        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        if readers != 0 || writers != 0 {
+            this.eval_libc_i32("EBUSY")
+        } else {
+            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
+            Ok(0)
+        }
+    }
+
+    fn pthread_rwlock_unlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
+        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        if let Some(new_readers) = readers.checked_sub(1) {
+            rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
+            Ok(0)
+        } else if writers != 0 {
+            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
+            Ok(0)
+        } else {
+            this.eval_libc_i32("EPERM")
+        }
+    }
+
+    fn pthread_rwlock_destroy(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        if rwlock_get_readers(this, rwlock_op)?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+        if rwlock_get_writers(this, rwlock_op)?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
```

Comment:
The [docs say](https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_rwlock_destroy.html):

>  Results are undefined if pthread_rwlock_destroy() is called when any thread holds rwlock.

So I think it would be more accurate to `throw_ub_format!` here.

Reply:
Generally, we want the shim to be cause UB whenever the specs allow it. This ensures that clients are confirming to any possible implementation.